src/sched/futex.c

```diff
@@ -274,62 +268,45 @@ int enclave_futex(
 {
     int rc;
     static int init = 1;
-    uint32_t bitset = FUTEX_BITSET_MATCH_ANY;
 
     if (init)
     {
         memset((void*)&futex_q_lock, 0, sizeof(struct ticketlock));
         init = 0;
     }
 
-    /* Ignore FUTEX_PRIVATE. We are single-process anyway. */
-    op &= ~(FUTEX_PRIVATE);
+    ticket_lock(&futex_q_lock);
 
-    if ((op & FUTEX_CLOCK_REALTIME) && !(op & (FUTEX_WAIT | FUTEX_WAIT_BITSET)))
+    if (op ==FUTEX_WAIT)
     {
-        return -ENOSYS;
-    }
-
-    clock_t clock =
-        op & FUTEX_CLOCK_REALTIME ? CLOCK_REALTIME : CLOCK_MONOTONIC;
-    op &= ~(FUTEX_CLOCK_REALTIME);
+        assert(lthread_self());
 
-    ticket_lock(&futex_q_lock);
-    switch (op)
+        rc = futex_wait(uaddr, val, timeout);
+        if (rc == 0 || rc == -ETIMEDOUT)
+        {
+            // return without unlocking
+            // we already unlocked when yielding in
+            // __do_futex_sleep
+            // N.B. the locking code here makes a lot
+            // of assumptions about code elsewhere and
+            // if you are looking for a concurrency bug,
+            // check this and make sure return value
+            // assumptions here haven't been changed
+            // in any code that futex_wait executes.
+            return rc;
+        }
+    }
+    else if (op == FUTEX_WAKE)
     {
-        case FUTEX_WAIT_BITSET:
-            if (val3 == 0)
-            {
-                rc = -EINVAL;
-                break;
-            }
-            bitset = (uint32_t)val3;
-            /* Fall through to FUTEX_WAIT */
-        case FUTEX_WAIT:
-            assert(lthread_self());
-
-            rc = futex_wait(uaddr, val, bitset, timeout, clock);
-            if (rc == 0 || rc == -ETIMEDOUT)
-                goto ret_nounlock;
-            break;
-        case FUTEX_WAKE_BITSET:
-            if (val3 == 0)
-            {
-                rc = -EINVAL;
-                break;
-            }
-            bitset = (uint32_t)val3;
-            /* Fall through to FUTEX_WAKE */
-        case FUTEX_WAKE:
-            rc = futex_wake(uaddr, val, bitset);
-            break;
-        default:
-            FUTEX_SGXLKL_VERBOSE("futex invalid op: %d\n", op);
-            rc = -ENOSYS;
+        rc = futex_wake(uaddr, val);
+    }
+    else
+    {
+        FUTEX_SGXLKL_VERBOSE("futex invalid op: %d\n", op);
```

Comment:
Let's just `sgxlkl_fail` here.  We tried to do a futex op that we shouldn't be doing.  Bad us!  That avoids some errors from things failing to check the return value of this.

Reply:
I'll do another PR to make that change.